apps/zotonic_mod_translation/src/support/translation_translate.erl

```diff
@@ -0,0 +1,154 @@
+    Texts :: [ Text ] | Text,
+    Text :: binary() | #trans{},
+    Context :: z:context(),
+    Translations :: [ binary() | undefined ],
+    Reason :: term().
+translate(FromLanguage, ToLanguage, Texts, Context) ->
+    FromCode = to_lang(FromLanguage, Context),
+    ToCode = to_lang(ToLanguage, Context),
+    translate_1(FromCode, ToCode, to_list(Texts), Context).
+
+translate_1({error, _}, _ToCode, _Texts, _Context) ->
+    {error, from_language};
+translate_1(_FromCode, {error, _}, _Texts, _Context) ->
+    {error, to_language};
+translate_1({ok, FromCode}, {ok, ToCode}, Texts, Context) ->
+    PartlyTranslated = lists:map(
+        fun(Text) ->
+            local_trans(FromCode, ToCode, trim(Text), Context)
+        end,
+        Texts),
+    add_service_trans(FromCode, ToCode, PartlyTranslated, Context).
+
+
+add_service_trans(FromCode, ToCode, PartlyTranslated, Context) ->
+    case lists:filtermap(
+        fun
+            ({FromText, undefined}) -> {true, FromText};
+            (_) -> false
+        end,
+        PartlyTranslated)
+    of
+        [] ->
+            Translations = [ T || {_, T} <- PartlyTranslated ],
+            {ok, Translations};
+        ToTranslate ->
+            Notification = #translate{
+                from = FromCode,
+                to = ToCode,
+                texts = ToTranslate
+            },
+            case z_notifier:first(Notification, Context) of
+                {ok, Translations} ->
+                    {ok, merge_translations(PartlyTranslated, Translations, [])};
+                {error, _} = Error ->
+                    Error;
+                undefined ->
+                    {_, Translations} = lists:unzip(PartlyTranslated),
+                    {ok, Translations}
+            end
+    end.
+
+merge_translations([], [], Acc) ->
+    lists:reverse(Acc);
+merge_translations([{_Text, undefined}|Rest], [Tr|Trs], Acc) ->
+    merge_translations(Rest, Trs, [ Tr | Acc ]);
+merge_translations([{_Text, Tr}|Rest], Trs, Acc) ->
+    merge_translations(Rest, Trs, [ Tr | Acc ]).
+
+
+local_trans(FromCode, ToCode, Text, _Context) when is_binary(Text), FromCode =:= ToCode ->
+    {Text, Text};
+local_trans(_FromCode, _ToCode, <<>>, _Context) ->
+    {<<>>, <<>>};
+local_trans(_FromCode, _ToCode, #trans{ tr = [] }, _Context) ->
+    {<<>>, <<>>};
+local_trans(FromCode, ToCode, #trans{ tr = Tr }, Context) ->
+    FromCodeText = lists:keyfind(FromCode, 1, Tr),
+    case lists:keyfind(ToCode, 1, Tr) of
+        {_ToCode, ToText} ->
+            {<<>>, ToText};
+        false when FromCode =:= en, FromCodeText =:= false ->
+            {<<>>, <<>>};
+        false when FromCode =:= en ->
+            {en, FromText} = FromCodeText,
+            case trans(FromText, ToCode, Context) of
+                undefined -> {FromText, undefined};
+                ToText -> {FromText, ToText}
+            end;
+        false when FromCodeText =:= false ->
+            {<<>>, <<>>};
```

Comment:
If this case is moved up, replacing `when FromCode =:= en, FromCodeText =:= false` the latter is not needed anymore - or am I missing some nuance?

Reply:
Also I don't understand yet why the result should be {<<>>, <<>>} - is that the encoding of a special case where actually the whole entry should be ignored? Shouldn't we encode that with `undefined` or `false` for filterMapping?

Reply:
FromCode doesn't need to be `en`.  Is it just that _if_ FromCode is `en` then we can use our `.po` translations to check for a known (and preferred) translation of the text.

That is why `en` is handled separately.

But indeed the `false when FromCodeText =:= false ->` can move up.